src/sdk/pynni/nni/compression/torch/pruning/structured_pruning.py

```diff
@@ -108,7 +111,156 @@ def get_mask(self, base_mask, weight, num_prune, wrapper, wrapper_idx):
         dict
             dictionary for storing masks
         """
-        raise NotImplementedError('{} get_mask is not implemented'.format(self.__class__.__name__))
+        raise NotImplementedError(
+            '{} get_mask is not implemented'.format(self.__class__.__name__))
+
+
+class ConstrainedStructuredWeightMasker(WeightMasker):
+    def calc_mask(self, wrappers, channel_dsets, groups, wrappers_idx):
+        """
+        Calculate the masks for the layers in the same dependency sets.
+        Similar to the traditional StructuredWeightMasker, ConstrainedStructuredWeightMasker
+        will prune the target layers based on the L1/L2 norm of the weights.
+        However, StructuredWeightMasker prunes the filter completely based on the
+        L1/L2 norm of each filter. In contrast, ConstrainedStructuredWeightMasker
+        will try to satisfy the channel/group dependency(see nni.compression.torch.
+        utils.shape_dependency for details). Specifically, ConstrainedStructuredWeightMasker
+        will try to prune the same channels for the layers that have channel dependency.
+        In addition, this mask calculator will also ensure that the number of filters
+        pruned in each group is the same(meet the group dependency).
+
+        Parameters
+        ----------
+        sparsities : list
+            List of float that specify the sparsity for each conv layer.
+        wrappers : list
+            List of wrappers
+        groups : list
+            The number of the filter groups of each layer.
+        wrappers_idx : list
+            The indexes of the wrappers
+        """
+        # sparsity configs for each wrapper
+        sparsities = [_w.config['sparsity'] for _w in wrappers]
+        # check the type of the input wrappers
+        for _w in wrappers:
+            msg = 'module type {} is not supported!'.format(_w.type)
+            assert _w.type == 'Conv2d', msg
```

Comment:
is there `Linear` here?

Reply:
Nope, these maskers are for FilterPruning, I think Linear is not included.

Reply:
in the implementation of channel dependency, you also add `Linear` into dependency group